### ISSUE TITLE
Error when passing list of callback functions to mesolve

### ISSUE THREAD

**Issue by BenjaminDAnjou** [qutip/qutip#2189]
### Bug Description

The documentation of mesolve states that lists of callback functions can be passed in the parameter e_ops:

> A list of operators as Qobj and/or callable functions (can be mixed) or a single callable function. For operators, the result’s expect will be computed by [qutip.expect](https://qutip.org/docs/latest/apidoc/functions.html#module-qutip.expect). For callable functions, they are called as f(t, state) and return the expectation value. A single callback’s expectation value can be any type, but a callback part of a list must return a number as the expectation value.

However, I find that passing a callback function in a list e_ops = [my_callback] gives an exception:

> UnboundLocalError: local variable 'rho_t' referenced before assignment

By contrast, I find that passing a callback function directly as e_ops = my_callback does not cause an issue.

This is important because I am unable to pass lists of callback functions which, according to the documentation, I should be able to do.

### Code to Reproduce the Bug

```shell
# Imports 
import qutip as qt
import numpy as np

# System parameters
w = 1
kappa = 0.1
# Size of the Hilbert space
n_max = 10
# Annihilation operator
a = qt.destroy(n_max)
# Callback function
my_callback = lambda t, rho_t : rho_t.purity()

# Hamiltonian
H = w * a.dag() * a
# Initial state
psi_0 = ( qt.fock(n_max, 0) + qt.fock(n_max, 1) ) / np.sqrt(2)
rho_0 = qt.ket2dm( psi_0)
# Integration times
times = np.linspace(0, 2/kappa, 5000)
# Collapse operators
c_ops = [np.sqrt(kappa) * a]
# Observables to compute
e_ops = [my_callback]

# Solve the master equation
output = qt.mesolve(H, rho_0, times, c_ops, e_ops)
```

### Code Output

```shell
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Cell In[17], line 29
     26 e_ops = [my_callback]
     28 # Solve the master equation
---> 29 output = qt.mesolve(H, rho_0, times, c_ops, e_ops)
     31 plt.plot(output.expect)

File ~/anaconda3/envs/danjou/lib/python3.10/site-packages/qutip/mesolve.py:244, in mesolve(H, rho0, tlist, c_ops, e_ops, args, options, progress_bar, _safe_mode)
    241     v = rho0.full().ravel('F')
    242     func(0., v, *ode_args) + v
--> 244 res = _generic_ode_solve(func, ode_args, rho0, tlist, e_ops, options,
    245                          progress_bar, dims=rho0.dims)
    246 res.num_collapse = len(c_ops)
    248 if e_ops_dict:

File ~/anaconda3/envs/danjou/lib/python3.10/site-packages/qutip/mesolve.py:519, in _generic_ode_solve(func, ode_args, rho0, tlist, e_ops, opt, progress_bar, dims)
    517 for m in range(n_expt_op):
    518     if not isinstance(e_ops[m], Qobj) and callable(e_ops[m]):
--> 519         output.expect[m][t_idx] = e_ops[m](t, rho_t)
    520         continue
    521     output.expect[m][t_idx] = expect_rho_vec(e_ops_data[m], r.y,
    522                                              e_ops[m].isherm
    523                                              and rho0.isherm)

UnboundLocalError: local variable 'rho_t' referenced before assignment
```

### Expected Behaviour

I expect that callback functions passed as a list should be computed without issue.

### Your Environment

```shell
QuTiP: Quantum Toolbox in Python
================================
Copyright (c) QuTiP team 2011 and later.
Current admin team: Alexander Pitchford, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, Eric Giguère, Boxi Li, Jake Lishman, Simon Cross and Asier Galicia.
Board members: Daniel Burgarth, Robert Johansson, Anton F. Kockum, Franco Nori and Will Zeng.
Original developers: R. J. Johansson & P. D. Nation.
Previous lead developers: Chris Granade & A. Grimsmo.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.7.1
Numpy Version:      1.24.3
Scipy Version:      1.10.1
Cython Version:     0.29.33
Matplotlib Version: 3.7.1
Python Version:     3.10.6
Number of CPUs:     12
BLAS Info:          OPENBLAS
OPENMP Installed:   False
INTEL MKL Ext:      False
Platform Info:      Darwin (arm64)
Installation path:  /Users/danb2901/anaconda3/envs/danjou/lib/python3.10/site-packages/qutip
================================================================================
Please cite QuTiP in your publication.
================================================================================
For your convenience a bibtex reference can be easily generated using `qutip.cite()`
```

**Comment by hodgestar** [comment]
This looks like a duplicate of #2118. A fix was just released in QuTiP 4.7.2 -- would you mind trying with that and letting us know if the fix works for you?

**Comment by BenjaminDAnjou** [comment]
> This looks like a duplicate of #2118. A fix was just released in QuTiP 4.7.2 -- would you mind trying with that and letting us know if the fix works for you?

Yup, this sure fixes it. Sorry for missing the duplicate!

**Comment by hodgestar** [comment]
No worries. Thank you for checking and reporting back.
